src/proposals/20250403/SparkEthereum_20250403.t.sol

```diff
@@ -0,0 +1,536 @@
+        assertEq(buidl.balanceOf(address(ctx.proxy)), 0);
+    }
+
+    function test_ETHEREUM_superstateUSTBOnboarding() public onChain(ChainIdUtils.Ethereum()) {
+        SparkLiquidityLayerContext memory ctx = _getSparkLiquidityLayerContext();
+
+        MainnetController controller = MainnetController(ETHEREUM_NEW_ALM_CONTROLLER);
+
+        IERC20 usdc           = IERC20(Ethereum.USDC);
+        ISuperstateToken ustb = ISuperstateToken(Ethereum.USTB);
+
+        bytes32 depositKey        = controller.LIMIT_SUPERSTATE_SUBSCRIBE();
+        bytes32 withdrawKey       = controller.LIMIT_SUPERSTATE_REDEEM();
+        bytes32 offchainRedeemKey = RateLimitHelpers.makeAssetDestinationKey(
+            controller.LIMIT_ASSET_TRANSFER(),
+            address(ustb),
+            address(ustb)
+        );
+
+        _assertRateLimit(depositKey,        0, 0);
+        _assertRateLimit(withdrawKey,       0, 0);
+        _assertRateLimit(offchainRedeemKey, 0, 0);
+
+        executeAllPayloadsAndBridges();
+
+        _assertRateLimit(depositKey, 300_000_000e6, 100_000_000e6 / uint256(1 days));
+        _assertRateLimit(withdrawKey, type(uint256).max, 0);
+        _assertRateLimit(offchainRedeemKey, type(uint256).max, 0);
+
+        // USDS -> USDC limits are 200m, go a bit below in case some is in use
+        uint256 mintAmount = 190_000_000e6;
+        vm.startPrank(ctx.relayer);
+        controller.mintUSDS(mintAmount * 1e12);
+        controller.swapUSDSToUSDC(mintAmount);
+
+        assertEq(usdc.balanceOf(address(ctx.proxy)), mintAmount);
+        assertEq(ustb.balanceOf(address(ctx.proxy)), 0);
+
+        assertEq(ctx.rateLimits.getCurrentRateLimit(depositKey), 300_000_000e6);
+
+        (uint256 ustbShares,,) = ustb.calculateSuperstateTokenOut(mintAmount, address(usdc));
+
+        controller.subscribeSuperstate(mintAmount);
+        vm.stopPrank();
+
+        assertEq(ctx.rateLimits.getCurrentRateLimit(depositKey), 300_000_000e6 - mintAmount);
+
+        assertEq(usdc.balanceOf(address(ctx.proxy)), 0);
+        assertEq(ustb.balanceOf(address(ctx.proxy)), ustbShares);
+
+        // Doing a smaller redeem because there is not necessarily enough liquidity
+        vm.prank(ctx.relayer);
+        controller.redeemSuperstate(ustbShares / 100);
+
+        assertApproxEqAbs(usdc.balanceOf(address(ctx.proxy)), mintAmount * 1/100, 100);
+        assertApproxEqAbs(ustb.balanceOf(address(ctx.proxy)), ustbShares * 99/100, 1);
+
+        // You can always burn the whole amount by doing it offchain
+        uint256 ustbBalance = ustb.balanceOf(address(ctx.proxy));
+        vm.prank(ctx.relayer);
+        controller.transferAsset(address(ustb), address(ustb), ustbBalance);
+
+        // USDC will come back async
+        assertApproxEqAbs(usdc.balanceOf(address(ctx.proxy)), mintAmount * 1/100, 100);
+        assertEq(ustb.balanceOf(address(ctx.proxy)), 0);
+    }
+
+    function test_ETHEREUM_centrifugeJTRSYOnboarding() public onChain(ChainIdUtils.Ethereum()) {
```

Comment:
Just adding some notes here for visibility:
1. @hieronx reviewed this test (specifically checking addresses/ids etc and `_centrifugeFulfillDepositRequest` and `_centrifugeFulfillRedeemRequest` helper functions and gave his sign off
1. The vault address used here is the correct one but it does not match the address used in testing in the v1.3.0 release of the spark-alm-controller repo. After talking with @hieronx we confirmed that this address was updated in the controller testing in this [PR](https://github.com/sparkdotfi/spark-alm-controller-private/pull/25). Its going in the v1.4.0 release.